packages/flutter/lib/src/gestures/events.dart

```diff
@@ -356,6 +434,61 @@ abstract class PointerEvent extends Diagnosticable {
   String toStringFull() {
     return toString(minLevel: DiagnosticLevel.fine);
   }
+
+  /// Returns the transformation of `position` into the coordinate system
+  /// described by `transform`.
+  ///
+  /// The z-value of `position` is assumed to be 0.0. If `transform` is null,
+  /// `position` is returned as-is.
+  static Offset transformPosition(Matrix4 transform, Offset position) {
+    if (transform == null) {
+      return position;
+    }
+    final Vector3 position3 = Vector3(position.dx, position.dy, 0.0);
+    final Vector3 transformed3 = transform.perspectiveTransform(position3);
+    return Offset(transformed3.x, transformed3.y);
+  }
+
+  /// Transforms `untransformedDelta` into the coordinate system described by
+  /// `transform`.
+  ///
+  /// It uses the provided `untransformedEndPosition` and
+  /// `transformedEndPosition` of the provided delta to increase accuracy.
+  ///
+  /// If `transform` is null, `untransformedDelta` is returned.
+  static Offset transformDeltaViaPositions({
+    @required Offset untransformedEndPosition,
+    Offset transformedEndPosition,
+    @required Offset untransformedDelta,
+    @required Matrix4 transform,
+  }) {
+    if (transform == null) {
+      return untransformedDelta;
+    }
+    // We could transform the delta directly with the transformation matrix.
+    // While that is mathematically equivalent, in practice we are seeing a
+    // greater precision error with that approach. Instead, we are transforming
+    // start and end point of the delta separately and calculate the delta in
+    // the new space for greater accuracy.
+    transformedEndPosition ??= transformPosition(transform, untransformedEndPosition);
+    final Offset transformedStartPosition = transformPosition(transform, untransformedEndPosition - untransformedDelta);
+    return transformedEndPosition - transformedStartPosition;
+  }
+
+  /// Takes a transform matrix meant for painting and returns a matrix that can
+  /// be used to transform [PointerEvent]s for hit testing.
+  ///
+  /// This method removes the "perspective" component from `paintTransform` by
+  /// setting the third column and third row of the matrix to "0, 0, 1, 0".
+  /// The resulting matrix can be used to determine the point in the local
+  /// coordinate system of the transformed component that is exactly under the
+  /// user's finger when the user is touching the screen.
+  static Matrix4 paintTransformToPointerEventTransform(Matrix4 paintTransform) {
```

Comment:
I don't understand why this would work without inverting the matrix. Let's say the paint transformation is to scale both x and y by a factor 2: `T = diag(2, 2, 1, 1)`. That means pixels of the widget are visually presented twice as far from the origin than in the local coordinate system. Then when a user taps the screen on one such pixel, that tap position will be twice as far from the origin than in the local coordinate system as well. So to do hit testing in the local coordinate system, the point would have to be transformed by `drag(0.5, 0.5, 1, 1)`. That should generalise to the inverse of `T`.

Reply:
I renamed this function to make it clearer what it actually does.